dfu: dfu_target: Reset a properly downloaded image on dfu_target_reset() (#9304)
nrfconnect/sdk-nrf: doc/nrf/libraries/dfu/dfu_target.rst

It then saves the data using buffered writes, resulting in fewer write operations and faster response times seen by the application.
The library also takes away, from the application, the responsibility of calculating the data-chunk write offset.

Common application flow
=======================

Each of the supported upgrade types follow the common DFU flow.

> 1. First initialize the specific DFU target by calling :c:func:`dfu_target_init`, internally it uses image type and image number to decide which target to initialize.
> 2. Ask the offset of already downloaded blocks by calling :c:func:`dfu_target_offset_get`. This step is optional, but it allows application to skip already downloaded blocks. Not all targets necessary supports it.
> 3. Write a block of DFU image into the target by calling :c:func:`dfu_target_write`. Repeat until all blocks have been downloaded.
> 4. When download is finnished, tell DFU library to that process is done by calling :c:func:`dfu_target_done`.
> 5. When application is ready to install the image, it can be marked to be updated by calling :c:func:`dfu_target_schedule_update`. Some targets do the installation then automatically on next boot.

Reviewer: ```suggestion
1. Initialize the specific DFU target by calling the :c:func:`dfu_target_init` function.

   Internally, it uses the image type and image number to decide which target to initialize.
#. Ask the offset of already downloaded blocks by calling the :c:func:`dfu_target_offset_get` function.

   This step is optional, but it allows the application to skip the blocks that are already downloaded.
   All targets do not necessarily support it.
#. Write a block of DFU image into the target by calling the :c:func:`dfu_target_write` function.

   Repeat until all blocks have been downloaded.
#. When all downloads have completed, call the :c:func:`dfu_target_done` function to tell the DFU library that the process has completed.
#. When the application is ready to install the image, call the :c:func:`dfu_target_schedule_update` function to mark it as ready for update.

   Some targets perform the installation automatically on next boot.
```

> To cancel ongoing operation call :c:func:`dfu_target_reset`, this should clear up any image that is already downloaded or even marked to be updated. This is different than aborting a download by just calling :c:func:`dfu_target_done`, in which case the same download could be resumed later.

Reviewer: ```suggestion
To cancel an ongoing operation, call the :c:func:`dfu_target_reset` function.
This should clear up any images that have already been downloaded or even marked to be updated.
This is different than aborting a download by calling the :c:func:`dfu_target_done` function, in which case the same download can be resumed later.
```

2. Call the :c:func:`dfu_target_schedule_update` function to mark the firmware as *ready to be booted*.
   On the next reboot, the device will run with the new firmware.

> When the application has to collect another MCUboot image, it must either erase the target area using :c:func:`dfu_target_reset` or schedule the upgrade for the current image.

Reviewer: ```suggestion
When the application has to collect another MCUboot image, it must either erase the target area using the :c:func:`dfu_target_reset` function or schedule the upgrade for the current image.
```

2. Call the :c:func:`dfu_target_schedule_update` function to mark the firmware as *ready to be booted*.
   On the next reboot, the device will run with the new firmware.

> When the application has to collect another MCUboot image, it must either erase the target area using the :c:func:`dfu_target_reset` function or schedule the upgrade for the current image.

Reviewer 2: Why does it have to do this? The update will never go through, its it from a usability perspective?

Author: To be honest, I don't know what the intend of this clause was originally. It sounds like it tries to describe two different use cases:
* If the application needs to overwrite the fota area with a new image, it must call `dfu_target_reset()`
* If the application wants to download another MCUboot image on another area, then it should just call `dfu_target_done()` and initialize the new target with a new call to `dfu_target_init()`

Author: I dropped the whole sentence.
This is more clearly expressed on the section above it, in `Common application flow`

The modem stores the data in the memory location reserved for firmware patches.
If there is already a firmware patch stored in the modem, the library requests the modem to delete the old firmware patch to make space for the new patch.

> When the transfer is completed, The application must call the :c:func:`dfu_target_done` function to release modem resources and then call :c:func:`dfu_target_schedule_update` to
> request the modem to apply the patch. On the next reboot, the modem will try to apply the patch.

Reviewer: ```suggestion
When the transfer has completed, the application must call the :c:func:`dfu_target_done` function to release modem resources and then call :c:func:`dfu_target_schedule_update` to request the modem to apply the patch.
On the next reboot, the modem tries to apply the patch.
```

> If existing image needs to be removed, even if it is marked to be updated, application may call :c:func:`dfu_target_reset` which will erase the DFU area and prepares it for next download.

Reviewer: ```suggestion
If an existing image needs to be removed, even if it is marked to be updated, the application may call the :c:func:`dfu_target_reset` function, which erases the DFU area and prepares it for next download.
```

The serialized firmware file uses the :file:`.cbor` extension.

This DFU target downloads the serialized modem firmware to an external flash memory.

> Once the modem firmware has been downloaded, the application should use :ref:`lib_fmfu_fdev` to write the firmware to the modem.

Reviewer: should or must?

Author: Or may.. or can..

I tried to clarify that this library does not do the update, it is the `lib_fmfu_fdev` which does it. So application can use that.

> DFU target library does not do the upgrade and calling :c:func:`dfu_target_schedule_update` does not do anything.

Reviewer: ```suggestion
The DFU target library does not perform the upgrade and calling the :c:func:`dfu_target_schedule_update` function has no effect.
```
